packages/react-reconciler/src/ReactFiberScheduler.js

```diff
@@ -1277,6 +1287,11 @@ function renderRoot(root: FiberRoot, isYieldy: boolean): void {
       resetContextDependences();
       resetHooks();
 
+      // Reset in case completion throws.
+      // This is only used in DEV and when replaying is on.
+      const mayReplay = mayReplayFailedUnitOfWork;
+      mayReplayFailedUnitOfWork = true;
```

Comment:
I don't love this boolean, especially because it's not dev only. But I'm more concerned that this replay logic is getting really convoluted and it seems there must be a better way to structure it.

Maybe we can move where `replayUnitOfWork` is called instead. If `replayUnitOfWork` is only called after a failed begin phase, we could fork `beginWork` to `beginWorkInDEV` which has a local try/catch.

```js
function beginWorkInDEV(current, workInProgress, renderExpirationTime) {
  try {
    beginWork(current, workInProgress, renderExpirationTime);
  } catch (e) {
    resetContextDependences();
    resetHooks();

    // Either call replayUnitOfWork or inline the whole thing here

    // Then throw to outer catch block
    throw e;
  }
}
```

There might be other implications that I'm forgetting, but I think we should give this a shot before piling more complexity onto this code path.

Reply:
>especially because it's not dev only

It is though. I think it would get DCE’d because we never read it outside DEV. The only reason I write to it outside a flag is just because it seems awkward to do extra wrapping because of scoping.

Reply:
Just checked — this is what I originally did, but @sebmarkbage asked that I move the extra `try` out of the hot path: https://github.com/facebook/react/pull/12201#discussion_r170399852

I feel we should reconsider given the additional complexity of distinguishing between the begin and complete phases.

Reply:
Pushed another commit that makes it clearer that part is DEV-only too. I'll look at your suggestion tomorrow/